Mathlib/AlgebraicGeometry/Morphisms/UniversallyInjective.lean

```diff
@@ -0,0 +1,95 @@
+/-
+Copyright (c) 2024 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.AlgebraicGeometry.PullbackCarrier
+import Mathlib.Topology.LocalAtTarget
+
+/-!
+# Universally closed morphism
+
+A morphism of schemes `f : X ⟶ Y` is universally closed if `X ×[Y] Y' ⟶ Y'` is injective
+for all base change `Y' ⟶ Y`. This is equivalent to the diagonal morphism being surjective
+(`AlgebraicGeometry.UniversallyInjective.iff_diagonal`).
+
+We show that being universally injective is local at the target, and is stable under
+compositions and base changes.
+
+## TODO
+- https://stacks.math.columbia.edu/tag/01S4
+  Show that this is equivalent to radicial morphisms
+  (injective + purely inseparable residue field extensions)
+
+-/
+
+noncomputable section
+
+open CategoryTheory CategoryTheory.Limits Opposite TopologicalSpace
+
+universe v u
+
+namespace AlgebraicGeometry
+
+variable {X Y : Scheme.{u}} (f : X ⟶ Y)
+
+open CategoryTheory.MorphismProperty Function
+
+/--
+A morphism of schemes `f : X ⟶ Y` is universally injective if the base change `X ×[Y] Y' ⟶ Y'`
+along any morphism `Y' ⟶ Y` is injective (on points).
+-/
+@[mk_iff]
+class UniversallyInjective (f : X ⟶ Y) : Prop where
+  out : universally (topologically (Injective ·)) f
```

Comment:
Is there a rule on fields named `out`? In this case, the class name is descriptive enough I suppose.

Reply:
I've renamed it.